libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -77,15 +82,46 @@ public SpanContext getCurrentSpan() {
         return (currentSpan == null) ? null : new SpanContext(currentSpan);
     }
 
-    private void endSpan(Span span) {
-        if (span != null) {
-            span.endSpan();
-            setCurrentSpanInContext(span.getParentSpan());
-        }
+    @Override
+    public ScopedSpan startScopedSpan(SpanCreationContext spanCreationContext) {
+        return startScopedSpan(spanCreationContext, null);
+    }
+
+    @Override
+    public ScopedSpan startScopedSpan(SpanCreationContext spanCreationContext, SpanContext parentSpan) {
+        Span span = startSpan(spanCreationContext.getSpanName(), parentSpan, spanCreationContext.getAttributes());
+        SpanScope spanScope = withSpanInScope(span);
+        return new DefaultScopedSpan(span, spanScope);
+    }
+
+    @Override
+    public SpanScope withSpanInScope(Span span) {
+        return DefaultSpanScope.create(span, tracerContextStorage).attach();
     }
 
     private Span createSpan(String spanName, Span parentSpan, Attributes attributes) {
-        return tracingTelemetry.createSpan(spanName, parentSpan, attributes);
+        return tracingTelemetry.createSpan(spanName, parentSpan, attributes, new SpanLifecycleListener() {
+            @Override
+            public void onStart(Span span) {
+                setCurrentSpanInContext(span);
+            }
+
+            @Override
+            public void onEnd(Span span) {
+                onSpanEnd(span, parentSpan);
+            }
+        });
+    }
+
+    private void onSpanEnd(Span span, Span parentSpan) {
```

Comment:
Do we still need to do anything on span end? I think we've been offloaded this to SpanScope  attach / detach?

Reply:
No I think, we can remove it. Should we remove the SpanLifecycleListener all together?

Reply:
Yes, please!